NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +52,224 @@ function Base.show(io::IO, s::CategoryProduct)
   return print(io, ")")
 end
 
-category_show(io::IO, k, v) = print(io, v)
-
+category_show(io::IO, ::Int, v) = print(io, v)
 category_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
 
+function Base.isless(s1::C, s2::C) where {C<:CategoryProduct}
+  return categories_isless(categories(s1), categories(s2))
+end
+function Base.isless(s1::CategoryProduct, s2::CategoryProduct)
+  return categories_isless(categories(s1), categories(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for CategoryProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+# get clean results when mixing implementations
+categories_isequal(::Tuple, ::NamedTuple) = false
+categories_isequal(::NamedTuple, ::Tuple) = false
```

Comment:
What about the empty case? Is that handled somewhere else?

I.e. I would expect `categories_isequal((), (;))` to be `true`.

Also, I might expect `categories_isequal((), (; A=U1(0)))` to be `true` (i.e. if the representation is the trivial representation). Curious to hear your and @emstoudenmire thoughts on those kinds of corner cases, maybe those cases are handled somewhere else in the code.

Reply:
As `CategoryProduct(::NamedTuple{()})` returns a `CategoryProduct(::Tuple{})`, the empty case cannot be a `NamedTuple`. Avoiding these edge case was one of the motivation for merging empty tuple and empty named tuple. Would you prefer using `EmptyCategoryStyle` to impose equality of all forms of `EmptyCategory` in a more generic way? 

`categories_isequal((), (; A=U1(0)))` is  handled by calling `sym_categories_insert_unspecified( (;), (A=U1(0))` (line 244) and returns true.

Reply:
I think maybe we should reassess how we are handling empty category products/empty categories in the code, it seems a little complicated right now.

Reply:
I think I'm following the logic of this now, and this design seems reasonable. Could you move the definitions:
```julia
function categories_isequal(nt::NamedTuple, ::Tuple{})
  return categories_isequal(nt, (;))
end
function categories_isequal(::Tuple{}, nt::NamedTuple)
  return categories_isequal((;), nt)
end
```
to this part of the code so it is clearer how those corner cases are handled? Part of my confusion was that definitions like `categories_isequal(::Tuple, ::NamedTuple) = false` might be wrong in the empty category product limit without those extra definitions.

Another question is, are there additional definitions like:
```julia
function categories_isequal(::NamedTuple{()}, t::Tuple)
  return categories_isequal((), t)
end
```
that handle empty NamedTuple, or are you assuming cases like that don't occur because `CategoryProduct((;))` converts to `CategoryProduct(())`?

I have to say that defining `CategoryProduct(::NamedTuple{()}) = CategoryProduct(())` doesn't sit right with me, I don't like the idea of doing a "bate and switch" on users and not returning an object based on the types they input. If the main implication of removing that definition is that we need to add a few more definitions for functions like `categories_isequal` that handle the empty `NamedTuple` limit, I would prefer that route.

I'm still a little suspicious of how empty category products are handled overall in this PR but we'll have to assess that on a case-by-case basis.

Reply:
I allowed `CategoryProduct{NamedTuple{()}}` again. Currently `CategoryProduct(SU2(1), U1(0)) == CategoryProduct(SU2(1))` is false, i.e. in the tuple case categories need to exactly match to be equal. Therefore CategoryProduct((;)) == CategoryProduct(U1(0))`` is false follows the logic and there is no need for dedicated code.

Reply:
My gut is telling me we should change `CategoryProduct(SU2(1), U1(0)) == CategoryProduct(SU2(1))` to return `true`. The main reason is that we want a system that is "open" in the sense of mixing different spaces together, and importantly these spaces will be made separately without any knowledge of the others. 

The main example would be making something like the Kondo lattice model: we would want "spin sites" with SU(2) only to be able to be mixed with Electron sites which would have SU2 and U1.

Reply:
I made the change such that the Tuple implementation also inserts trivial sectors inside unspecified ones.